src/zookeeper.ts

```diff
@@ -1,316 +1,432 @@
-    if (err) {
-      emitter.emit(LocatorException.CODE["FAILED_TO_GET_CHILDREN"], path, err);
-      setPool([]);
-      processQueue();
-      return;
+    /**
+     * Connect to the zookeeper cluster, rebuilding the client on session expired.
+     */
+    public connect() {
+        this.serverLocator()
+            .then((location) => {
+                var zookeeperServer = location.host;
+                if (location.port) {
+                    zookeeperServer = zookeeperServer + ":" + location.port;
+                }
+                zookeeperServer = zookeeperServer + this.discoveryPath;
+
+                if (this.client) {
+                    this.client.removeAllListeners();
+                }
+
+                this.client = zookeeper.createClient(zookeeperServer, {
+                    sessionTimeout: this.sessionTimeout,
+                    spinDelay: this.spinDelay,
+                    retries: this.retries
+                });
+
+                this.clientWrapper.setClient(this.client);
+
+                var disconnectedHandler = null;
+
+                this.client.on("connected", () => {
+                    this.emit(LocatorException.CODE.CONNECTED);
+
+                    if (disconnectedHandler) {
+                        clearTimeout(disconnectedHandler);
+                    }
+                });
+                this.client.on("disconnected", (count : number) => {
+                    this.emit(LocatorException.CODE.DISCONNECTED);
+
+                    disconnectedHandler = setTimeout(
+                        () => {
+                            this.client.close();
+
+                            // further delaying the emit of 'expired' event seems to avoid
+                            // intermittent errors thrown by node-zookeeper-client
+                            setTimeout(() => {
+                                this.client.emit("expired");
+                            }, 1000);
+                        }
+                        , this.sessionTimeout);
+                });
+                this.client.on('state', (state : Object) => {
+                    this.emit(LocatorException.CODE.STATE_CHANGE, state, this.client.getSessionId());
+                });
+                this.client.on("expired", () => {
+                    this.emit(LocatorException.CODE.EXPIRED);
+                    this.connect();
+                });
+
+                this.emit(LocatorException.CODE.CONNECTING);
+                this.client.connect();
+            })
+            .catch((err) => {
+                this.emit(LocatorException.CODE.ZK_LOCATOR_ERROR, err);
+                this.connect();
+            });
     }
+}
 
-    var promises = children.map(function (child) {
-      var deferred = Promise.defer<Location>();
+/**
+ * ZookeeperPathLocatorFactory provides Locators for a list of hosts stored in a path in a zookeeper cluster
+ */
+class ZookeeperPathLocatorFactory {
```

Comment:
There are currently 4 classes, 2 interfaces and a function in this file, with 1 of the classes, `ZookeeperLocator` and 1 of the interfaces, `ZookeeperLocatorParameters` exported. At this point i sort of think it makes sense to leave it as is rather than split into 3+ files. It still weighs in under 500 lines, even with jsdoc comments.

Perhaps once additional service discovery mechanisms are added it would make sense to divide the src folder into sub folders for each locator type, and split this file at this point.